### PR TITLE
Introduce contribution guidelines and DCO check

### LINKED ISSUE #139: Require contributors to sign off commits
We want to ensure contributors to our repo have the right to submit their code. A lightweight way to achieve this is to require contributors to sign off commits, and agree to a Developer Certificate of Origin [https://developercertificate.org/].

## Definition of done
- We have a CONTRIBUTORS.md file stating that signed off commits agree to the [DCO](https://developercertificate.org/).
- Our CI fails on commits that do not conform to this.

### PR DESCRIPTION
**Pull request by netrome**
This PR introduces contribution guidelines in `CONTRIBUTING.md`. Among these guidelines are requirements to sign off of any commit. This is enforced in CI using [dco-check](https://github.com/christophebedard/dco-check).

Closes #139

### CODE REVIEW

**Comment by netrome** [comment]
Task failed successfully https://github.com/Trust-Machines/core-eng/actions/runs/4797508229/jobs/8534582095?pr=292

image
image

File: CODE_OF_CONDUCT.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+* Discrimination, harassment, or mistreatment based on race, gender, sexual orientation, religion, disability, age, or any other basis
+* Personal attacks, threats, or intimidation directed towards others
+* Inappropriate, offensive, or harmful language, images, or content
+* Disrespectful or dismissive behavior, including public or private harassment
+* Deliberate disruption of discussions, meetings, or collaborative work
+* Plagiarism, violation of intellectual property rights, or taking credit for others' work
+
+## 4. Reporting Violations
+
+If you witness or experience any violation of this Code of Conduct or have concerns about someone's behavior, please report the incident to [marten@trustmachines.co] or any other project maintainer. They will investigate and take appropriate action, respecting the confidentiality of the reporter.
```
[the remark this answers is not in the document]
Comment (netrome) [reply CODE_OF_CONDUCT.md:29]: Sure, perhaps `core-eng-leads@trustmachines.co` is appropriate. wdyt?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
+# Contributing to CoreEng
+
+Welocme to the CoreEng repo!
```
[the remark this answers is not in the document]
Comment (netrome) [reply CONTRIBUTING.md:3]: Hmm, this should be lowercased actually. I'll change to "core-eng".

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+2. **Provide a clear and concise title**: Write a meaningful title that reflects the issue's content. This helps maintainers and contributors quickly grasp the issue's context.
+3. **Add a definition of done**: Provide a clear criteria for when the issue is concidered done. This ensures that the implementation lives up to the original intention of the issue.
+4. **Include necessary details**: Explain the issue thoroughly, including steps to reproduce (if applicable), expected behavior, and encountered problems. Include relevant screenshots or log files to better illustrate the issue.
+5. **Stay on topic**: Keep the discussion focused on the issue at hand. Avoid discussing unrelated topics or providing unnecessary information.
+
+## Pull Requests
+
+When creating a pull request, please follow these guidelines to ensure a smooth and efficient review process:
+
+1. **Build on the main branch**: We don't want long-lived feature branches. All changes should be integrated into `main` as early as possible.
```
[the remark this answers is not in the document]
Comment (netrome) [reply CONTRIBUTING.md:26]: Hmm. Long-lived is a confusing term here since it's not about time. I don't care if it takes years for an external contributor to develop a feature, as long as it results in a small and focused PR. I'll rephrase this to state that we don't maintain other branches than `main`.

[the remark this answers is not in the document]
Comment (netrome) [reply CONTRIBUTING.md]: Updated.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
+# Contributing to CoreEng
+
+Welocme to the CoreEng repo!
+This repo is primarily used by the core engineering team at TrustMachines, but anyone is welcome to contribute.
+This document provides guidelines and instructions to help make the contribution process smooth and efficient.
+
+## Code of Conduct
+
+By participating in this project, you agree to abide by our [Code of Conduct](CODE_OF_CONDUCT.md).
+Please report any unacceptable behavior to [marten@trustmachines.co] or any other project maintainer.
```
[the remark this answers is not in the document]
Comment (netrome) [reply CONTRIBUTING.md]: Done.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
+# Contributing to CoreEng
+
+Welocme to the CoreEng repo!
+This repo is primarily used by the core engineering team at TrustMachines, but anyone is welcome to contribute.
```
[the remark this answers is not in the document]
Comment (netrome) [reply CONTRIBUTING.md]: Done.

File: CONTRIBUTING.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (netrome) [reply CONTRIBUTING.md]: Done.

File: CODE_OF_CONDUCT.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (netrome) [reply CODE_OF_CONDUCT.md]: Done.

**Review by stjepangolemac — DISMISSED** [review]

File: CODE_OF_CONDUCT.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (netrome) [reply CODE_OF_CONDUCT.md]: Updated.

File: CODE_OF_CONDUCT.md [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+* Discrimination, harassment, or mistreatment based on race, gender, sexual orientation, religion, disability, age, or any other basis
+* Personal attacks, threats, or intimidation directed towards others
+* Inappropriate, offensive, or harmful language, images, or content
+* Disrespectful or dismissive behavior, including public or private harassment
+* Deliberate disruption of discussions, meetings, or collaborative work
+* Plagiarism, violation of intellectual property rights, or taking credit for others' work
+
+## 4. Reporting Violations
+
+If you witness or experience any violation of this Code of Conduct or have concerns about someone's behavior, please report the incident to [core-eng-violations@trustmachines.co](mailto:core-eng-violations@trustmachines.co) or any project maintainer.
```
[the remark this answers is not in the document]
Comment (netrome) [reply CODE_OF_CONDUCT.md:29]: Yeah I commented in the other thread. I think core-eng-violations is a better name. I created it because I didn't have authorization to rename the existing group. Can you change this group to have the right people and authorization rights?

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
+# Contributing to core-eng
+
+Welocme to the core-eng repo!
+This repo is primarily used by the Core Engineering team at TrustMachines, but anyone is welcome to contribute.
+This document provides guidelines and instructions to help make the contribution process smooth and efficient.
+
+## Code of Conduct
+
+By participating in this project, you agree to abide by our [Code of Conduct](CODE_OF_CONDUCT.md).
+Please report any unacceptable behavior to [core-eng-violations@trustmachines.co](mailto:core-eng-violations@trustmachines.co) or any other project maintainer.
```
[the remark this answers is not in the document]
Comment (netrome) [reply CONTRIBUTING.md:10]: Awaiting your reply in the other thread before making any changes. I think this is a better name for the group.

File: CODE_OF_CONDUCT.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (netrome) [reply CODE_OF_CONDUCT.md:29]: Video chat: We agreed on the name `open-source-concerns@trustmachines.co`. I'll change to it.

[the remark this answers is not in the document]
Comment (netrome) [reply CODE_OF_CONDUCT.md]: Done.

File: CONTRIBUTING.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (netrome) [reply CONTRIBUTING.md]: Done.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,53 @@
+# Contributing to core-eng
+
+Welocme to the core-eng repo!
```
Comment (jferrant) [inline CONTRIBUTING.md:3]: Typo.

**Review by jferrant — APPROVED** [review]
Just the one typo that I saw. Looks good :D I better go link some issues to some PRs I have open....

**Comment by netrome** [comment]
> Just the one typo that I saw. Looks good :D I better go link some issues to some PRs I have open....

Thnaks for spotting! I don't know how this happened, my spelling is normally flawelss :sweat_smile:.

**Review by jferrant — APPROVED** [review]

**Comment by netrome** [comment]
> Some know if this works with GH Desktop too? I am using that visual app.

Since this is in CI, it doesn't matter how you create your commits. Use whatever process you like. This PR is frozen though, as we're awaiting some input before taking action.

**Comment by sergey-shandar** [comment]
@netrome Do you still need the branch?

**Comment by netrome** [comment]
> @netrome Do you still need the branch?

Nope, thanks for reminding.
